src/core/lib/event_engine/posix_engine/posix_engine.cc

```diff
@@ -636,14 +572,83 @@ EventEngine::ConnectionHandle PosixEventEngine::Connect(
 #endif  // GRPC_PLATFORM_SUPPORTS_POSIX_POLLING
 }
 
+EventEngine::ConnectionHandle PosixEventEngine::CreateEndpointFromUnconnectedFd(
+    int fd, EventEngine::OnConnectCallback on_connect,
+    const EventEngine::ResolvedAddress& addr, const EndpointConfig& config,
+    MemoryAllocator memory_allocator, EventEngine::Duration timeout) {
+#if GRPC_PLATFORM_SUPPORTS_POSIX_POLLING
+  int err;
+  int connect_errno;
+  do {
+    err = connect(fd, addr.address(), addr.size());
+  } while (err < 0 && errno == EINTR);
+  connect_errno = (err < 0) ? errno : 0;
+
+  auto addr_uri = ResolvedAddressToURI(addr);
+  if (!addr_uri.ok()) {
+    Run([on_connect = std::move(on_connect),
+         ep = absl::FailedPreconditionError(absl::StrCat(
+             "connect failed: ", "invalid addr: ",
+             addr_uri.value()))]() mutable { on_connect(std::move(ep)); });
+    return EventEngine::ConnectionHandle::kInvalid;
+  }
+
+  std::string name = absl::StrCat("tcp-client:", addr_uri.value());
+  PosixTcpOptions options = TcpOptionsFromEndpointConfig(config);
+  PosixEventPoller* poller = poller_manager_->Poller();
+  EventHandle* handle =
+      poller->CreateHandle(fd, name, poller->CanTrackErrors());
+
+  if (connect_errno == 0) {
+    // Connection already succeeded. Return 0 to discourage any cancellation
+    // attempts.
+    Run([on_connect = std::move(on_connect),
+         ep = CreatePosixEndpoint(
+             handle, nullptr, shared_from_this(), std::move(memory_allocator),
+             options)]() mutable { on_connect(std::move(ep)); });
+    return EventEngine::ConnectionHandle::kInvalid;
+  }
+  if (connect_errno != EWOULDBLOCK && connect_errno != EINPROGRESS) {
+    // Connection already failed. Return 0 to discourage any cancellation
+    // attempts.
+    handle->OrphanHandle(nullptr, nullptr, "tcp_client_connect_error");
+    Run([on_connect = std::move(on_connect),
+         ep = absl::FailedPreconditionError(absl::StrCat(
+             "connect failed: ", "addr: ", addr_uri.value(),
+             " error: ", std::strerror(connect_errno)))]() mutable {
+      on_connect(std::move(ep));
+    });
+    return EventEngine::ConnectionHandle::kInvalid;
+  }
+  // Connection is still in progress.
+  int64_t connection_id =
+      last_connection_id_.fetch_add(1, std::memory_order_acq_rel);
+  AsyncConnect* ac = new AsyncConnect(
+      std::move(on_connect), shared_from_this(), executor_.get(), handle,
+      std::move(memory_allocator), options, addr_uri.value(), connection_id);
+  int shard_number = connection_id % connection_shards_.size();
+  struct ConnectionShard* shard = &connection_shards_[shard_number];
+  {
+    grpc_core::MutexLock lock(&shard->mu);
+    shard->pending_connections.insert_or_assign(connection_id, ac);
+  }
+  // Start asynchronous connect and return the connection id.
+  ac->Start(timeout);
+  return {static_cast<intptr_t>(connection_id), 0};
+#else   // GRPC_PLATFORM_SUPPORTS_POSIX_POLLING
+  grpc_core::Crash(
```

Comment:
nit: rather than have 60 lines guarded by an `#if` block, you can add this crash in just 3 lines at the start of the function.

```
#ifndef GRPC_PLATFORM_SUPPORTS_POSIX_POLLING
grpc_core::Crash("EventEngine::CancelConnect is not supported on this platform");
#endif
```